src/upchunk.ts

```diff
@@ -305,37 +346,55 @@ export class UpChunk  {
     this.getChunk()
       .then(() => {
         this.attemptCount = this.attemptCount + 1;
-
+        this.lastChunkStart = new Date();
         return this.sendChunk()
       })
       .then((res) => {
         if (SUCCESSFUL_CHUNK_UPLOAD_CODES.includes(res.statusCode)) {
+          const lastChunkEnd = new Date();
+          const lastChunkInterval = (lastChunkEnd.getTime() - this.lastChunkStart.getTime()) / 1000;
+
           this.dispatch('chunkSuccess', {
             chunk: this.chunkCount,
+            chunkSize: this.chunkSize,
             attempts: this.attemptCount,
+            timeInterval: lastChunkInterval,
             response: res,
           });
 
           this.attemptCount = 0;
           this.chunkCount = this.chunkCount + 1;
+          this.nextChunkRangeStart = this.nextChunkRangeStart + this.chunkByteSize;
 
           if (this.chunkCount < this.totalChunks) {
+            // dynamic chunk sizing
+            if (this.dynamicChunkSize) {
+              if (lastChunkInterval < 10)
+              {
+                this.chunkSize = Math.min(this.chunkSize * 2,this.maxChunkSize);
+              } else if (lastChunkInterval > 30) {
+                this.chunkSize = Math.max(this.chunkSize / 2,this.minChunkSize);
```

Comment:
We force chunk sizes to be a multiple of 256kB, but this logic doesn't preserve that invariant.

Does anything actually rely on the chunk size being a multiple of 256kB?

Reply:
The GCP documentation states that the chunk size must be multiples of 256kB.  I'll improve the logic here.

Reply:
OK, fixed!